qt/aqt/mediasrv.py

```diff
@@ -28,146 +31,127 @@ def _getExportFolder():
+    try:
+        isdir = os.path.isdir(os.path.join(directory, path))
+    except ValueError:
+        return flask.Response(
+            "Path for '%s - %s' is too long!" % (directory, path),
+            status=HTTPStatus.BAD_REQUEST,
+            mimetype="text/plain",
+        )
 
-        ctype = self.guess_type(path)
-        try:
-            f = open(path, "rb")
-        except OSError:
-            self.send_error(HTTPStatus.NOT_FOUND, "File not found")
-            return None
-        try:
-            self.send_response(HTTPStatus.OK)
-            self.send_header("Content-type", ctype)
-            fs = os.fstat(f.fileno())
-            self.send_header("Content-Length", str(fs[6]))
-            self.send_header("Last-Modified", self.date_time_string(fs.st_mtime))
-            self.send_header("Access-Control-Allow-Origin", "*")
-            self.end_headers()
-            return f
-        except:
-            f.close()
-            raise
-
-    def log_message(self, format, *args):
-        if not devMode:
-            return
-        print(
-            "%s - - [%s] %s"
-            % (self.address_string(), self.log_date_time_string(), format % args)
+    if isdir:
+        return flask.Response(
+            "Path for '%s - %s' is a directory (not supported)!" % (directory, path),
+            status=HTTPStatus.FORBIDDEN,
+            mimetype="text/plain",
         )
 
-    def _redirectWebExports(self, path):
-        # catch /_anki references and rewrite them to web export folder
-        targetPath = os.path.join(os.getcwd(), "_anki", "")
-        if path.startswith(targetPath):
-            newPath = os.path.join(_exportFolder, path[len(targetPath) :])
-            return newPath
+    try:
+        if devMode:
+            print("Sending file '%s - %s'" % (directory, path))
+        return flask.send_from_directory(directory, path)
+
+    except Exception as error:
+        if devMode:
+            print(
+                "Caught HTTP server exception,\n%s"
+                % "".join(traceback.format_exception(*sys.exc_info())),
+            )
+
+        # swallow it - user likely surfed away from
+        # review screen before an image had finished
+        # downloading
+        return flask.Response(
+            "For path '%s - %s' %s!" % (directory, path, error),
+            status=HTTPStatus.INTERNAL_SERVER_ERROR,
+            mimetype="text/plain",
+        )
 
-        # catch /_addons references and rewrite them to addons folder
-        targetPath = os.path.join(os.getcwd(), "_addons", "")
-        if path.startswith(targetPath):
-            try:
-                addMgr = self.mw.addonManager
-            except AttributeError:
-                return path
 
-            addonPath = path[len(targetPath) :]
+def _redirectWebExports(path):
+    # catch /_anki references and rewrite them to web export folder
+    targetPath = "_anki/"
+    if path.startswith(targetPath):
```

Comment:
Is path already normalized? I'm wondering what happens in the case of /_anki/../../some/filesystem/path

Reply:
I tried doing this, but paths as `/../something` entries are converted to `/something`. I run this code on the `QTWEBENGINE_REMOTE_DEBUGGING=8080` console:
```
var request = new XMLHttpRequest();
request.open("GET", "/../deleted.txt", true);
request.send();
```

The result on the server was:
```
Sending file 'D:\User\Documents\Anki2\Evandro\collection.media - deleted.txt'
Caught HTTP server exception,
Traceback (most recent call last):
  File "F:\anki\qt\aqt\mediasrv.py", line 106, in allroutes
    return flask.send_from_directory(directory, path)
  File "F:\anki\pyenv\lib\site-packages\flask\helpers.py", line 767, in send_from_directory
    raise NotFound()
werkzeug.exceptions.NotFound: 404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
```

Also, the function `flask.send_from_directory` used, states that it is a safe function as long the base directory is correct. In our case, it will send anything inside `addMgr.addonsFolder()`, `_redirectWebExports.mw.col.media.dir()` and `_exportFolder`.